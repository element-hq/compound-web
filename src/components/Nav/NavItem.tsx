/*
Copyright 2023 New Vector Ltd.
Copyright 2022 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, {
  type AnchorHTMLAttributes,
  type ButtonHTMLAttributes,
  type ForwardedRef,
  type MouseEventHandler,
  forwardRef,
} from "react";

import styles from "./Nav.module.css";

type NavItemProps = {
  active?: boolean;
  "aria-controls"?: string;
};

type NavItemLinkProps = Omit<
  AnchorHTMLAttributes<HTMLAnchorElement>,
  "style" | "className"
> & {
  href: string;
} & NavItemProps;

type NavItemButtonProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  "style" | "className"
> & {
  onClick: MouseEventHandler<HTMLButtonElement>;
} & NavItemProps;

const NavItemLink = forwardRef(function NavItemLink(
  {
    children,
    href,
    onClick,
    ...rest
  }: React.PropsWithChildren<NavItemLinkProps>,
  ref: ForwardedRef<HTMLAnchorElement>,
) {
  return (
    <a
      href={href}
      onClick={onClick}
      className={styles["nav-item"]}
      {...rest}
      ref={ref}
    >
      {children}
    </a>
  );
});

const NavItemButton = forwardRef(function NavItemButton(
  {
    children,
    disabled,
    onClick,
    ...rest
  }: React.PropsWithChildren<NavItemButtonProps>,
  ref: ForwardedRef<HTMLButtonElement>,
) {
  return (
    <button
      onClick={onClick}
      className={styles["nav-item"]}
      disabled={disabled}
      ref={ref}
      {...rest}
    >
      {children}
    </button>
  );
});

const renderAsLink = (
  props: React.PropsWithChildren<NavItemLinkProps | NavItemButtonProps>,
): props is React.PropsWithChildren<NavItemLinkProps> =>
  "href" in props && !!props.href;

/**
 * A navigation item component to be used with a navigation bar.
 * Will render an anchor when href is provided, otherwise a button element.
 */
export const NavItem = forwardRef(function NavItem(
  props: React.PropsWithChildren<NavItemLinkProps | NavItemButtonProps>,
  ref:
    | ForwardedRef<HTMLButtonElement | null>
    | ForwardedRef<HTMLAnchorElement | null>,
) {
  /**
   * For accessibility rules related to tabs,
   * see https://www.w3.org/WAI/ARIA/apg/patterns/tabs/#wai-ariaroles,states,andproperties
   *
   * For accessibility rules related to navigation,
   * see https://www.digitala11y.com/navigation-role/
   */
  const ariaControls = props["aria-controls"];
  const isUsedAsTabs = !!ariaControls;
  const a11yAttributes = isUsedAsTabs
    ? {
        // when used as tabs
        "aria-controls": ariaControls,
        role: "tab",
        "aria-selected": props.active,
      }
    : {
        // when used as navigation elements
        "aria-current": props.active ? true : undefined,
      };

  // All the attributes except `active` can be passed to the button/a element.
  const propsForChild = { ...props };
  delete propsForChild["active"];

  // Depending on whether `href` is in props, we render link/button
  let navItem: React.ReactNode;
  if (renderAsLink(propsForChild)) {
    const buttonRef = ref as ForwardedRef<HTMLAnchorElement | null>;
    navItem = (
      <NavItemLink {...propsForChild} {...a11yAttributes} ref={buttonRef} />
    );
  } else {
    const buttonRef = ref as ForwardedRef<HTMLButtonElement | null>;
    navItem = (
      <NavItemButton {...propsForChild} {...a11yAttributes} ref={buttonRef} />
    );
  }

  return (
    <li
      className={styles["nav-tab"]}
      data-current={props.active ? true : undefined}
      role="presentation"
    >
      {navItem}
    </li>
  );
});
