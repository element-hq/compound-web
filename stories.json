{"v":3,"stories":{"actioncontrol--round":{"id":"actioncontrol--round","name":"Round","title":"ActionControl","importPath":"./src/components/ActionControl/ActionControl.stories.tsx","tags":["story"],"kind":"ActionControl","story":"Round","parameters":{"__id":"actioncontrol--round","docsOnly":false,"fileName":"./src/components/ActionControl/ActionControl.stories.tsx"}},"alert--docs":{"id":"alert--docs","title":"Alert","name":"Docs","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Alert","story":"Docs","parameters":{"__id":"alert--docs","docsOnly":true,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--success":{"id":"alert--success","name":"Success","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Success","parameters":{"__id":"alert--success","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--critical":{"id":"alert--critical","name":"Critical","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Critical","parameters":{"__id":"alert--critical","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--info":{"id":"alert--info","name":"Info","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Info","parameters":{"__id":"alert--info","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--without-close":{"id":"alert--without-close","name":"Without Close","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Without Close","parameters":{"__id":"alert--without-close","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"avatar--docs":{"id":"avatar--docs","title":"Avatar","name":"Docs","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Avatar","story":"Docs","parameters":{"__id":"avatar--docs","docsOnly":true,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--round":{"id":"avatar--round","name":"Round","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Round","parameters":{"__id":"avatar--round","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--square":{"id":"avatar--square","name":"Square","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Square","parameters":{"__id":"avatar--square","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--button":{"id":"avatar--button","name":"Button","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Button","parameters":{"__id":"avatar--button","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--no-image-fallback":{"id":"avatar--no-image-fallback","name":"No Image Fallback","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"No Image Fallback","parameters":{"__id":"avatar--no-image-fallback","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--large-no-image-fallback":{"id":"avatar--large-no-image-fallback","name":"Large No Image Fallback","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Large No Image Fallback","parameters":{"__id":"avatar--large-no-image-fallback","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--all-avatars":{"id":"avatar--all-avatars","name":"All Avatars","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"All Avatars","parameters":{"__id":"avatar--all-avatars","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar-stackedavatar--docs":{"id":"avatar-stackedavatar--docs","title":"Avatar/StackedAvatar","name":"Docs","importPath":"./src/components/Avatar/AvatarStack.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Avatar/StackedAvatar","story":"Docs","parameters":{"__id":"avatar-stackedavatar--docs","docsOnly":true,"fileName":"./src/components/Avatar/AvatarStack.stories.tsx"}},"avatar-stackedavatar--avatar-stack":{"id":"avatar-stackedavatar--avatar-stack","name":"Avatar Stack","title":"Avatar/StackedAvatar","importPath":"./src/components/Avatar/AvatarStack.stories.tsx","tags":["autodocs","story"],"kind":"Avatar/StackedAvatar","story":"Avatar Stack","parameters":{"__id":"avatar-stackedavatar--avatar-stack","docsOnly":false,"fileName":"./src/components/Avatar/AvatarStack.stories.tsx"}},"badge--badge":{"id":"badge--badge","name":"Badge","title":"Badge","importPath":"./src/components/Badge/Badge.stories.tsx","tags":["story"],"kind":"Badge","story":"Badge","parameters":{"__id":"badge--badge","docsOnly":false,"fileName":"./src/components/Badge/Badge.stories.tsx"}},"button--docs":{"id":"button--docs","title":"Button","name":"Docs","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Button","story":"Docs","parameters":{"__id":"button--docs","docsOnly":true,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--primary":{"id":"button--primary","name":"Primary","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Primary","parameters":{"__id":"button--primary","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--secondary":{"id":"button--secondary","name":"Secondary","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Secondary","parameters":{"__id":"button--secondary","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--tertiary":{"id":"button--tertiary","name":"Tertiary","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Tertiary","parameters":{"__id":"button--tertiary","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--destructive":{"id":"button--destructive","name":"Destructive","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Destructive","parameters":{"__id":"button--destructive","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"checkbox--docs":{"id":"checkbox--docs","title":"Checkbox","name":"Docs","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Checkbox","story":"Docs","parameters":{"__id":"checkbox--docs","docsOnly":true,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--primary":{"id":"checkbox--primary","name":"Primary","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Primary","parameters":{"__id":"checkbox--primary","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--critical":{"id":"checkbox--critical","name":"Critical","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Critical","parameters":{"__id":"checkbox--critical","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--disabled":{"id":"checkbox--disabled","name":"Disabled","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Disabled","parameters":{"__id":"checkbox--disabled","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--disabled-checked":{"id":"checkbox--disabled-checked","name":"Disabled Checked","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Disabled Checked","parameters":{"__id":"checkbox--disabled-checked","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"form--root":{"id":"form--root","name":"Root","title":"Form","importPath":"./src/components/Form/Form.stories.tsx","tags":["story"],"kind":"Form","story":"Root","parameters":{"__id":"form--root","docsOnly":false,"fileName":"./src/components/Form/Form.stories.tsx"}},"glass--docs":{"id":"glass--docs","title":"Glass","name":"Docs","importPath":"./src/components/Glass/Glass.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Glass","story":"Docs","parameters":{"__id":"glass--docs","docsOnly":true,"fileName":"./src/components/Glass/Glass.stories.tsx"}},"glass--glass":{"id":"glass--glass","name":"Glass","title":"Glass","importPath":"./src/components/Glass/Glass.stories.tsx","tags":["autodocs","story"],"kind":"Glass","story":"Glass","parameters":{"__id":"glass--glass","docsOnly":false,"fileName":"./src/components/Glass/Glass.stories.tsx"}},"icon--icon":{"id":"icon--icon","name":"Icon","title":"Icon","importPath":"./src/components/Icon/Icon.stories.tsx","tags":["story"],"kind":"Icon","story":"Icon","parameters":{"__id":"icon--icon","docsOnly":false,"fileName":"./src/components/Icon/Icon.stories.tsx"}},"button-iconbutton--docs":{"id":"button-iconbutton--docs","title":"Button/IconButton","name":"Docs","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Button/IconButton","story":"Docs","parameters":{"__id":"button-iconbutton--docs","docsOnly":true,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"button-iconbutton--normal":{"id":"button-iconbutton--normal","name":"Normal","title":"Button/IconButton","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","story"],"kind":"Button/IconButton","story":"Normal","parameters":{"__id":"button-iconbutton--normal","docsOnly":false,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"button-iconbutton--with-indicator":{"id":"button-iconbutton--with-indicator","name":"With Indicator","title":"Button/IconButton","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","story"],"kind":"Button/IconButton","story":"With Indicator","parameters":{"__id":"button-iconbutton--with-indicator","docsOnly":false,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"button-iconbutton--with-highlight-indicator":{"id":"button-iconbutton--with-highlight-indicator","name":"With Highlight Indicator","title":"Button/IconButton","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","story"],"kind":"Button/IconButton","story":"With Highlight Indicator","parameters":{"__id":"button-iconbutton--with-highlight-indicator","docsOnly":false,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"link--docs":{"id":"link--docs","title":"Link","name":"Docs","importPath":"./src/components/Link/Link.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Link","story":"Docs","parameters":{"__id":"link--docs","docsOnly":true,"fileName":"./src/components/Link/Link.stories.tsx"}},"link--round":{"id":"link--round","name":"Round","title":"Link","importPath":"./src/components/Link/Link.stories.tsx","tags":["autodocs","story"],"kind":"Link","story":"Round","parameters":{"__id":"link--round","docsOnly":false,"fileName":"./src/components/Link/Link.stories.tsx"}},"menuitem--docs":{"id":"menuitem--docs","title":"MenuItem","name":"Docs","importPath":"./src/components/MenuItem/MenuItem.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"MenuItem","story":"Docs","parameters":{"__id":"menuitem--docs","docsOnly":true,"fileName":"./src/components/MenuItem/MenuItem.stories.tsx"}},"menuitem--primary":{"id":"menuitem--primary","name":"Primary","title":"MenuItem","importPath":"./src/components/MenuItem/MenuItem.stories.tsx","tags":["autodocs","story"],"kind":"MenuItem","story":"Primary","parameters":{"__id":"menuitem--primary","docsOnly":false,"fileName":"./src/components/MenuItem/MenuItem.stories.tsx"}},"menuitem--critical":{"id":"menuitem--critical","name":"Critical","title":"MenuItem","importPath":"./src/components/MenuItem/MenuItem.stories.tsx","tags":["autodocs","story"],"kind":"MenuItem","story":"Critical","parameters":{"__id":"menuitem--critical","docsOnly":false,"fileName":"./src/components/MenuItem/MenuItem.stories.tsx"}},"togglemenuitem--docs":{"id":"togglemenuitem--docs","title":"ToggleMenuItem","name":"Docs","importPath":"./src/components/MenuItem/ToggleMenuItem.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"ToggleMenuItem","story":"Docs","parameters":{"__id":"togglemenuitem--docs","docsOnly":true,"fileName":"./src/components/MenuItem/ToggleMenuItem.stories.tsx"}},"togglemenuitem--primary":{"id":"togglemenuitem--primary","name":"Primary","title":"ToggleMenuItem","importPath":"./src/components/MenuItem/ToggleMenuItem.stories.tsx","tags":["autodocs","story"],"kind":"ToggleMenuItem","story":"Primary","parameters":{"__id":"togglemenuitem--primary","docsOnly":false,"fileName":"./src/components/MenuItem/ToggleMenuItem.stories.tsx"}},"radio--docs":{"id":"radio--docs","title":"Radio","name":"Docs","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Radio","story":"Docs","parameters":{"__id":"radio--docs","docsOnly":true,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--primary":{"id":"radio--primary","name":"Primary","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Primary","parameters":{"__id":"radio--primary","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--checked":{"id":"radio--checked","name":"Checked","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Checked","parameters":{"__id":"radio--checked","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--checked-disabled":{"id":"radio--checked-disabled","name":"Checked Disabled","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Checked Disabled","parameters":{"__id":"radio--checked-disabled","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--critical":{"id":"radio--critical","name":"Critical","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Critical","parameters":{"__id":"radio--critical","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--critical-checked":{"id":"radio--critical-checked","name":"Critical Checked","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Critical Checked","parameters":{"__id":"radio--critical-checked","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--critical-checked-disabled":{"id":"radio--critical-checked-disabled","name":"Critical Checked Disabled","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Critical Checked Disabled","parameters":{"__id":"radio--critical-checked-disabled","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"search--docs":{"id":"search--docs","title":"Search","name":"Docs","importPath":"./src/components/Search/Search.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Search","story":"Docs","parameters":{"__id":"search--docs","docsOnly":true,"fileName":"./src/components/Search/Search.stories.tsx"}},"search--search":{"id":"search--search","name":"Search","title":"Search","importPath":"./src/components/Search/Search.stories.tsx","tags":["autodocs","story"],"kind":"Search","story":"Search","parameters":{"__id":"search--search","docsOnly":false,"fileName":"./src/components/Search/Search.stories.tsx"}},"toggle--docs":{"id":"toggle--docs","title":"Toggle","name":"Docs","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Toggle","story":"Docs","parameters":{"__id":"toggle--docs","docsOnly":true,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--active":{"id":"toggle--active","name":"Active","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Active","parameters":{"__id":"toggle--active","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--checked":{"id":"toggle--checked","name":"Checked","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Checked","parameters":{"__id":"toggle--checked","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--disabled":{"id":"toggle--disabled","name":"Disabled","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Disabled","parameters":{"__id":"toggle--disabled","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--disabled-checked":{"id":"toggle--disabled-checked","name":"Disabled Checked","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Disabled Checked","parameters":{"__id":"toggle--disabled-checked","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"tooltip--docs":{"id":"tooltip--docs","title":"Tooltip","name":"Docs","importPath":"./src/components/Tooltip/Tooltip.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Tooltip","story":"Docs","parameters":{"__id":"tooltip--docs","docsOnly":true,"fileName":"./src/components/Tooltip/Tooltip.stories.tsx"}},"tooltip--side":{"id":"tooltip--side","name":"Side","title":"Tooltip","importPath":"./src/components/Tooltip/Tooltip.stories.tsx","tags":["autodocs","story"],"kind":"Tooltip","story":"Side","parameters":{"__id":"tooltip--side","docsOnly":false,"fileName":"./src/components/Tooltip/Tooltip.stories.tsx"}},"tooltip--align":{"id":"tooltip--align","name":"Align","title":"Tooltip","importPath":"./src/components/Tooltip/Tooltip.stories.tsx","tags":["autodocs","story"],"kind":"Tooltip","story":"Align","parameters":{"__id":"tooltip--align","docsOnly":false,"fileName":"./src/components/Tooltip/Tooltip.stories.tsx"}},"typography--heading":{"id":"typography--heading","name":"Heading","title":"Typography","importPath":"./src/components/Typography/Heading.stories.tsx","tags":["story"],"kind":"Typography","story":"Heading","parameters":{"__id":"typography--heading","docsOnly":false,"fileName":"./src/components/Typography/Heading.stories.tsx"}},"typography--text":{"id":"typography--text","name":"Text","title":"Typography","importPath":"./src/components/Typography/Text.stories.tsx","tags":["story"],"kind":"Typography","story":"Text","parameters":{"__id":"typography--text","docsOnly":false,"fileName":"./src/components/Typography/Text.stories.tsx"}}}}
