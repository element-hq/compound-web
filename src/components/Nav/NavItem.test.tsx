/*
Copyright 2023 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import React from "react";
import { composeStories } from "@storybook/react";

import * as stories from "./NavItem.stories";

const { Default, Disabled, Link, Active, ActiveLink, ActiveDisabled } =
  composeStories(stories);

describe("<NavItem />", () => {
  it("render a default item", () => {
    const { container } = render(<Default />);
    expect(container).toMatchSnapshot();
  });

  it("render a Disabled item", () => {
    const { container } = render(<Disabled />);
    expect(container).toMatchSnapshot();
  });
  it("render a Link item", () => {
    const { container } = render(<Link />);
    expect(container).toMatchSnapshot();
  });
  it("render a Active item", () => {
    const { container } = render(<Active />);
    expect(container).toMatchSnapshot();
  });
  it("render a ActiveLink item", () => {
    const { container } = render(<ActiveLink />);
    expect(container).toMatchSnapshot();
  });
  it("render a ActiveDisabled item", () => {
    const { container } = render(<ActiveDisabled />);
    expect(container).toMatchSnapshot();
  });
});
