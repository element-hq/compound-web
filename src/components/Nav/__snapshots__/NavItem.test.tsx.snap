// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<NavItem /> > render a Active item 1`] = `
<div>
  <nav
    aria-label="Testing"
    class="_nav-bar_0471db"
    role="navigation"
  >
    <ul
      class="_nav-bar-items_0471db"
    >
      <li
        class="_nav-tab_0471db"
        data-current="true"
        role="presentation"
      >
        <button
          aria-current="true"
          class="_nav-item_0471db"
        >
          Sessions
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<NavItem /> > render a ActiveDisabled item 1`] = `
<div>
  <nav
    aria-label="Testing"
    class="_nav-bar_0471db"
    role="navigation"
  >
    <ul
      class="_nav-bar-items_0471db"
    >
      <li
        class="_nav-tab_0471db"
        data-current="true"
        role="presentation"
      >
        <button
          aria-current="true"
          class="_nav-item_0471db"
          disabled=""
        >
          Sessions
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<NavItem /> > render a ActiveLink item 1`] = `
<div>
  <nav
    aria-label="Testing"
    class="_nav-bar_0471db"
    role="navigation"
  >
    <ul
      class="_nav-bar-items_0471db"
    >
      <li
        class="_nav-tab_0471db"
        data-current="true"
        role="presentation"
      >
        <a
          aria-current="true"
          class="_nav-item_0471db"
          href="https://example.org"
        >
          Sessions
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<NavItem /> > render a Disabled item 1`] = `
<div>
  <nav
    aria-label="Testing"
    class="_nav-bar_0471db"
    role="navigation"
  >
    <ul
      class="_nav-bar-items_0471db"
    >
      <li
        class="_nav-tab_0471db"
        role="presentation"
      >
        <button
          class="_nav-item_0471db"
          disabled=""
        >
          Sessions
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<NavItem /> > render a Link item 1`] = `
<div>
  <nav
    aria-label="Testing"
    class="_nav-bar_0471db"
    role="navigation"
  >
    <ul
      class="_nav-bar-items_0471db"
    >
      <li
        class="_nav-tab_0471db"
        role="presentation"
      >
        <a
          class="_nav-item_0471db"
          href="https://example.org"
        >
          Sessions
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<NavItem /> > render a default item 1`] = `
<div>
  <nav
    aria-label="Testing"
    class="_nav-bar_0471db"
    role="navigation"
  >
    <ul
      class="_nav-bar-items_0471db"
    >
      <li
        class="_nav-tab_0471db"
        role="presentation"
      >
        <button
          class="_nav-item_0471db"
        >
          Sessions
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
